sips/sip-clarity4/sip-clarity4.md

```diff
@@ -0,0 +1,466 @@
+built-in keyword `current-contract` is introduced. This keyword provides a
+straightforward way to access the current contract's principal.
+
+- **Description**: Returns the principal of the current contract.
+- **Output**: `principal`
+- **Example**:
+  ```clarity
+  (stx-transfer? u1000000 tx-sender current-contract)
+  ```
+
+## Conversion to `string-ascii`: `to-ascii?`
+
+Originally proposed here: https://github.com/clarity-lang/reference/issues/82
+
+`to-ascii?` is a new Clarity function that converts simple values into their
+`string-ascii` representations.
+
+- **Input**: `int` | `uint` | `bool` | `principal` | `(buff 524284)` |
+  `(string-utf8 262144)`
+- **Output**: `(response (string-ascii 1048571) uint)`
+- **Signature**: `(to-ascii? value)`
+- **Description**: Returns the `string-ascii` representation of the input value
+  in an `ok` response on success. The only error condition is if the input type
+  is `string-utf8` and the value contains non-ASCII characters, in which case,
+  `(err u1)` is returned.
+- **Example**:
+  ```clarity
+  (to-ascii? true) ;; Returns (ok "true")
+  (to-ascii? 42) ;; Returns (ok "42")
+  (to-ascii? 'SP2QEZ06AGJ3RKJPBV14SY1V5BBFNAW33D96YPGZF) ;; Returns (ok "SP2QEZ06AGJ3RKJPBV14SY1V5BBFNAW33D96YPGZF")
+  (to-ascii? 0x12345678) ;; Returns (ok "0x12345678")
+  ```
+
+## Timestamp for current block: `block-time`
+
+`block-time` is a new Clarity keyword that returns the timestamp of the current
+block in seconds since the Unix epoch. This is the same timestamp that is in the
+block header, which is verified by the signers to be:
+
+- Greater than the timestamp of its parent block
+- At most 15 seconds into the future
+
+This same timestamp can also be retrieved for previous blocks using
+`(get-stacks-block-info? time height)`, which exists since Clarity 3, but cannot
+be used for the current block.
+
+Note that `block-time` will properly account for the context of an `at-block`
+expression. If the `at-block` sets the context to a block that is from before
+Clarity 4 has activated, attempting to use `block-time` in that context will
+result in a runtime error.
+
+- **Output**: `uint`
+- **Example**:
+  ```clarity
+  (if (> block-time 1755820800)
+    (print "after 2025-07-22")
+    (print "before 2025-07-22"))
+  ```
+
+# Related Work
+
+Not applicable.
+
+# Backwards Compatibility
+
+Because this SIP introduces new Clarity operators, it is a consensus-breaking
+change. A contract that uses one of these new operators would be invalid before
+this SIP is activated, and valid after it is activated.
+
+The old `as-contract` builtin function is no longer available in Clarity 4,
+since it has been replaced with the safer `as-contract?`.
+
+All new keywords introduced in Clarity 4 can no longer be used as identifiers in
+a Clarity 4 smart contract. This means that some existing contracts would need
+to be modified if published as Clarity 4 contracts.
+
```

Comment:
Expand with migration considerations. What tools or warnings will guide developers upgrading contracts or using older (clarity 1, 2, 3) contracts?

It may be obvious for people who have worked with Clarity for longer but not for newer users. How to use `at block` or Clarity 1, 2 and 3 contracts after Clarity 4 is live.

Reply:
I added one sentence here, but I'm not sure what else is needed.